### PR TITLE
[#1936] Migrate store.js to TypeScript

### PR DESCRIPTION
**Pull request by vvidday**
Part of #1936

### Proposed commit message

```
Currently, despite the addition of TypeScript support, the frontend is
still largely written in JavaScript. This results in a lack of type
safety and many complex objects being passed around as unknown types,
which may in turn lead to errors.

Let's migrate the store.js file to TypeScript, to get types for the all
the states that are stored within the Vuex store. This will provide a
good starting point to migrate the Vue files that access and mutate the
state of the store.
```

### Other information

The declaration of the module and `ComponentCustomProperties` in `vuex.d.ts` provides type definitions to the `this.$store` object that's used in the `.vue` files.

Also, although I have checked on my end, this PR should probably not be merged until after #1913 , so we can see the updated results of the CI that includes the new TypeScript rules.

### CODE REVIEW

File: frontend/src/types/vuex.d.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  summaryDates: SummaryDates;
+  mergedGroups: string[]
+  fileTypeColors: { [key: string]: string }
+  loadingOverlayCount: number
+  loadingOverlayMessage: string
+  isTabActive: boolean
+}
+
+declare module '@vue/runtime-core' {
+  interface State {
```
Comment (ckcherry23) [inline frontend/src/types/vuex.d.ts:51]: Would it be possible to reuse the `StoreState` interface here?

**Review by ckcherry23 — COMMENTED** [review]
Thanks for your work! I just have a few questions about typing.

**Review by ckcherry23 — APPROVED** [review]
LGTM!

File: frontend/src/store/store.ts [hunk 2]
```diff
@@ -1,55 +1,62 @@
 import { createStore } from 'vuex';
+import { AuthorshipFile, CommitResult, DailyCommit } from '../types/types';
+import {
+  AuthorshipInfo,
+  StoreState,
+  SummaryDates,
+  ZoomInfo,
+} from '../types/vuex.d';
 
 export default createStore({
```
[the remark this answers is not in the document]
Comment (vvidday) [reply frontend/src/store/store.ts]: Thanks for catching this!

File: frontend/src/store/store.ts [hunk 3]
```diff
@@ -1,55 +1,62 @@
...
     mergedGroups: [],
     fileTypeColors: {},
     loadingOverlayCount: 0,
     loadingOverlayMessage: '',
     isTabActive: true,
-  },
+  } as StoreState,
   mutations: {
-    updateTabZoomInfo(state, info) {
+    updateTabZoomInfo(state, info: ZoomInfo) {
```
[the remark this answers is not in the document]
Comment (vvidday) [reply frontend/src/store/store.ts]: Great point, I've added in the types for the `state` variable - I think the `context` is already typed

File: frontend/src/types/vuex.d.ts (hunk 1 above)
Comment (vvidday) [reply frontend/src/types/vuex.d.ts]: Pushed a commit for the state to extend the `StoreState` interface, thanks for pointing this out!

File: frontend/src/types/vuex.d.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+interface SummaryDates {
+  since: string;
+  until: string;
+}
+
+interface StoreState {
+  tabAuthorshipInfo: AuthorshipInfo;
+  tabZoomInfo: ZoomInfo;
+  summaryDates: SummaryDates;
+  mergedGroups: string[]
```
Comment (HCY123902) [inline frontend/src/types/vuex.d.ts:43]: Just a comment regarding format consistency. I think the semicolons are missing here. Actually, the current eslint rules do not seem to enforce the presence of semicolons, which can be discussed in a separate issue

**Review by HCY123902 — CHANGES_REQUESTED** [review]

File: frontend/src/types/vuex.d.ts (hunk 4 above)
Comment (vvidday) [reply frontend/src/types/vuex.d.ts]: Thanks for pointing this out! Just fixed the semicolons in the latest commits.

Just checked the ESLint rules that we're adding in #1912 - the ruleset does enforce semicolons for regular statements (the [semi](https://typescript-eslint.io/rules/semi/) rule), but not for TypeScript `interface`s and `type`s specifically, as you mentioned. 

We can enforce this via the [member-delimiter-style](https://typescript-eslint.io/rules/member-delimiter-style) rule, i.e. just adding the following to `.eslintrc.json`.
```
"rules": {
  "@typescript-eslint/member-delimiter-style": "error"
}
```
Will open up a separate issue about this where we can investigate rules like the above one. Thanks!

**Review by HCY123902 — APPROVED** [review]

**Comment by vvidday** [comment]
Just added a ESLint exception in https://github.com/reposense/RepoSense/pull/1937/commits/c010d83a038e6037bec1405182eb91d0b0b84f79 for the empty interface in preparation for #1912 to be merged - empty interfaces are usually not meaningful, but this case is an exception.

Because the empty/"repeated" interface is inside the `declare module` block, which is used to type the `this.$store` within Vue components. We can't export to other files from within the `declare module` block, hence the need to declare `StoreState` at the top level (for other TS files), and once more in the `declare module` block for the Vue-component-specific typings.

**Review by zhoukerrr — APPROVED** [review]
LGTM! @ckcherry23 caught the points that I want to bring up. Thanks for the help!

**Pull request merged**